### PR TITLE
Fix torrent files prioritization

### PR DESCRIPTION
**Pull request by d-komarov**
After setting file priority, any subsequent attempt to set different
priorities will fail if there is a `file_priority` job running in in
disk thread. This happens because `torrent::m_file_priority` is being
updated before adding disk thread job. The problem is gone if the file
priority vector owned by the torrent object, is modified in the
`torrent::on_file_priority` callback, when disk job finishes.

* Fix torrent test
* Minor `torrent` class refactoring

### CODE REVIEW

File: include/libtorrent/torrent.hpp [hunk 1]
```diff
@@ -1182,6 +1181,7 @@ namespace libtorrent
 		bool request_bandwidth_from_session(int channel) const;
 
 		void update_peer_interest(bool was_finished);
+		void update_piece_priorities();
```
Comment (d-komarov) [inline include/libtorrent/torrent.hpp]: Moved to private section

File: bindings/python/test.py [hunk 2]
```diff
@@ -50,6 +50,8 @@ def test_torrent_handle(self):
 		self.assertEqual(self.h.piece_priorities(), [4])
 
 		self.h.prioritize_files([0,1])
+		# workaround for asynchronous priority update
+		time.sleep(1)
```
Comment (arvidn) [inline bindings/python/test.py]: good find. I thought the priorities were still updated immediately, but then corrected when the disk job come back, is that not the case?

File: src/torrent.cpp [hunk 3]
```diff
@@ -5664,99 +5675,66 @@ namespace {
...
+				boost::bind(&torrent::on_file_priority, shared_from_this(), _1));
 		}
-
-		update_piece_priorities();
 	}
 
 	void torrent::set_file_priority(int index, int prio)
 	{
-		INVARIANT_CHECK;
+		if (index < 0 || (valid_metadata() && index >= m_torrent_file->num_files())) { return; }
```
Comment (arvidn) [inline src/torrent.cpp]: I would like to preserve the formatting style. and was there something wrong with the comment?
also, can you comment on why the invariant check and check for `is_seed()` was removed?

File: bindings/python/test.py (hunk 2 above)
Comment (arvidn) [reply bindings/python/test.py]: I would prefer to restore the previous behavior here

File: src/torrent.cpp [hunk 4]
```diff
@@ -1142,17 +1142,13 @@ namespace libtorrent
...
-		// in share mode, all pieces have their priorities initialized to 0
-		if (m_share_mode && valid_metadata())
+		if (m_share_mode)
 		{
-			m_file_priority.clear();
-			m_file_priority.resize(m_torrent_file->num_files(), 0);
+			int const num_files = valid_metadata() ? m_torrent_file->num_files() : m_file_priority.size();
+			// in share mode, all pieces have their priorities initialized to 0
+			prioritize_files(std::vector<int>(num_files));
+			recalc_share_mode();
```
Comment (arvidn) [inline src/torrent.cpp]: this seems like a reasonable refactor, but unrelated to the issue, right?

File: src/torrent.cpp [hunk 5]
```diff
@@ -2022,8 +2018,7 @@ namespace libtorrent
 
 		// in case file priorities were passed in via the add_torrent_params
 		// and also in the case of share mode, we need to update the priorities
-		if (!m_file_priority.empty() && std::find(m_file_priority.begin()
-				, m_file_priority.end(), 0) != m_file_priority.end())
+		if (std::find(m_file_priority.begin(), m_file_priority.end(), 0) != m_file_priority.end())
```
Comment (arvidn) [inline src/torrent.cpp]: I would prefer to keep lines shorter than this

File: src/torrent.cpp [hunk 6]
```diff
@@ -5664,99 +5675,66 @@ namespace {
...
 		boost::scoped_ptr<std::vector<boost::uint8_t> > p(j->buffer.priorities);
-		if (m_file_priority == *p) return;
 
-		// in this case, some file priorities failed to get set
-		m_file_priority = *p;
-		update_piece_priorities();
+		if (j->ret == piece_manager::no_error)
+		{
+			TORRENT_ASSERT(!j->error);
+			m_file_priority = *p;
```
Comment (arvidn) [inline src/torrent.cpp]: do we even need to assign the priorities in this case?

File: src/torrent.cpp [hunk 7]
```diff
@@ -5664,99 +5675,66 @@ namespace {
 	void torrent::on_file_priority(disk_io_job const* j)
```
Comment (arvidn) [inline src/torrent.cpp]: is this the actual fix?

File: src/torrent.cpp [hunk 8]
```diff
@@ -5664,99 +5675,66 @@ namespace {
...
+			return;
 		}
 
+		std::vector<boost::uint8_t> const new_priority = fix_priorities(files, m_torrent_file->files());
+
+		if (!m_torrent_file->num_pieces() || is_seed() || new_priority == m_file_priority) { return; }
+
 		// storage may be NULL during shutdown
-		if (m_torrent_file->num_pieces() > 0 && m_storage)
+		if (boost::shared_ptr<piece_manager> const tmp = m_storage)
```
Comment (arvidn) [inline src/torrent.cpp]: why do you need to make a copy of the storage pointer? surely `m_storage` won't get destroyed from under our feet here, before we issue the disk call

**Comment by arvidn** [comment]
would you mind splitting these changes up into smaller, independent, pieces?
It would make it much simpler to see which changes belong together.

Does this patch make updating of priorities asynchronous? or did I do that in my patch? (I would prefer them not to be).

Perhaps a better solution would be for the disk job to send back just the diff of priorities. Or maybe we would only overwrite the file priorities if there's a disk error, and they deviate.

my understanding is that the fundamental case that's failing is:

1. set priorities -> issues async disk job in the background
2. set priorities to something else
3. the first disk job comes back and overwrites the priorities from (2)

File: src/torrent.cpp (hunk 3 above)
Comment (d-komarov) [reply src/torrent.cpp]: This method is only about to construct priority vector and pass it to `torrent::prioritize_files` that has all checks preserved. Logic for adding disk job was also removed from here.

File: src/torrent.cpp (hunk 5 above)
Comment (d-komarov) [reply src/torrent.cpp]: I'm not aware of current line length limitation so I assumed to align this to the longest line(# 2061) of the function.

File: src/torrent.cpp (hunk 6 above)
Comment (d-komarov) [reply src/torrent.cpp]: This change actually fixes the issue.

File: src/torrent.cpp (hunk 7 above)
Comment (d-komarov) [reply src/torrent.cpp]: Yes, it is in this function, also in `torrent::prioritize_files` and `torrent::set_file_priority`.

File: src/torrent.cpp (hunk 8 above)
Comment (d-komarov) [reply src/torrent.cpp]: I'll overview lifetime management of storage object again, to give the correct answer.
But at the moment I was writing this, I thought that it could be possible to destroy storage from other thread by external plugins.

File: src/torrent.cpp (hunk 4 above)
Comment (d-komarov) [reply src/torrent.cpp]: Each piece of code that deals with file priorities is presumably related.

**Comment by d-komarov** [comment]
I'll try to handle this, but it'll take some time, because I'm going to leave for a week.

Yes, this patch makes the file priorities change asynchronous.

How that priority diff will be used? I'm not sure what solution is better. Need to consider both cons and pros to make up a decision.

Your understanding is correct.

File: bindings/python/test.py (hunk 2 above)
Comment (d-komarov) [reply bindings/python/test.py]: I've added this to get python tests passed. This workaround depends on `torrent::prioritize_files` implementation that is considered in this PR.

File: src/torrent.cpp (hunk 5 above)
Comment (d-komarov) [reply src/torrent.cpp]: Creating actual '.clang-format' file may be a good idea.

File: src/torrent.cpp (hunk 4 above)
Comment (arvidn) [reply src/torrent.cpp]: so the relevant change here is to call `prioritize_files()` regardless of whether we have metadata or not, right?

File: src/torrent.cpp (hunk 5 above)
Comment (arvidn) [reply src/torrent.cpp]: yes, I've considered that. I haven't had time to learn how to describe the existing formatting rules in libtorrent though. I didn't realise you had change this line other than reformatting it. This is also not a functional change, right?

File: src/torrent.cpp [hunk 9]
```diff
@@ -5664,99 +5675,66 @@ namespace {
...
-		// initialize pad files to priority 0
-		file_storage const& fs = m_torrent_file->files();
-		for (int i = 0; i < (std::min)(fs.num_files(), limit); ++i)
+		// store file priorities until we get the metadata
+		if (!valid_metadata())
 		{
-			if (!fs.pad_file_at(i)) continue;
-			m_file_priority[i] = 0;
+			m_file_priority.assign(files.begin(), files.end());
+			return;
```
Comment (arvidn) [inline src/torrent.cpp]: I see, you're saving off the priorities here, so you can apply them immediately when we receive the metadata. is that right? (if so, please explain that in a comment)

File: src/torrent.cpp [hunk 10]
```diff
@@ -5664,99 +5675,66 @@ namespace {
...
 		}
 
+		std::vector<boost::uint8_t> const new_priority = fix_priorities(files, m_torrent_file->files());
+
+		if (!m_torrent_file->num_pieces() || is_seed() || new_priority == m_file_priority) { return; }
```
Comment (arvidn) [inline src/torrent.cpp]: it seems reasonable to move the checks that are independent of `new_priority` before doing the work of `fix_priorites()`, no?

File: src/torrent.cpp (hunk 8 above)
Comment (arvidn) [reply src/torrent.cpp]: `m_storage` is only supposed to be changed from the main (network) thread. most of libtorrent is single threaded in a sense. The exceptions are the `alert_manager` (which can be called directly from the user thread) and the disk threads (which is communicated with via disk jobs and post results back to the main thread, via the `io_service`)

File: bindings/python/test.py (hunk 2 above)
Comment (arvidn) [reply bindings/python/test.py]: maybe you're right. It does make sense to apply file priorities asynchronously, since the alternative is to (potentially) lie about the priorities. If we can't apply it, we shouldn't ever say it was applied.

The problem with making it async though is that there really should be an alert to indicate that it was applied.

File: src/torrent.cpp (hunk 6 above)
Comment (arvidn) [reply src/torrent.cpp]: I think we need to assign this unconditionally here. (i.e. hoist it out of the if-block). Even if there is an error reported from the disk thread, some priorities may have been applied.

File: src/torrent.cpp (hunk 9 above)
Comment (arvidn) [reply src/torrent.cpp]: oh, I see you did leave a comment

File: src/torrent.cpp (hunk 3 above)
Comment (arvidn) [reply src/torrent.cpp]: please expand the `{ return; }` block to match the existing formatting, or remove the `{` and `}`

File: src/torrent.cpp [hunk 11]
```diff
@@ -5784,17 +5762,11 @@ namespace {
 	{
 		INVARIANT_CHECK;
 
-		if (!valid_metadata())
-		{
-			files->resize(m_file_priority.size());
-			std::copy(m_file_priority.begin(), m_file_priority.end(), files->begin());
-			return;
-		}
+		files->assign(m_file_priority.begin(), m_file_priority.end());
```
Comment (arvidn) [inline src/torrent.cpp]: this new code is not quite equivalent. previously the size of `files` was preserved. Now it isn't. Was that deliberate?

File: test/test_torrent.cpp [hunk 12]
```diff
@@ -52,110 +52,117 @@ POSSIBILITY OF SUCH DAMAGE.
...
+			if (h.file_priorities() == prio) { return true; }
 
-	std::vector<boost::uint8_t> zeroes;
-	zeroes.resize(1000, 0);
-	add_torrent_params p;
-	p.flags &= ~add_torrent_params::flag_paused;
-	p.flags &= ~add_torrent_params::flag_auto_managed;
-	p.ti = info;
-	p.save_path = ".";
+			test_sleep(100);
```
Comment (arvidn) [inline test/test_torrent.cpp]: ideally this should be waiting for an alert. but I'm not sure it's possible to add a new alert without risking ABI incompatibilities. maybe it has to be added to `master`

**Comment by arvidn** [comment]
I think your approach of setting file priorities async. is better. It makes more sense compared to the alternatives

File: bindings/python/test.py (hunk 2 above)
Comment (d-komarov) [reply bindings/python/test.py]: Yes, I like the idea of introducing new alert for this operation. This what I thought of, when posted an issue for improvement.

File: src/torrent.cpp (hunk 3 above)
Comment (d-komarov) [reply src/torrent.cpp]: No problem.

File: src/torrent.cpp (hunk 4 above)
Comment (d-komarov) [reply src/torrent.cpp]: Yes, I believe that priorities must be handled in one place, that is `torrent::prioritize_files`. 'Empty metadata' case is covered there.

File: src/torrent.cpp (hunk 5 above)
Comment (d-komarov) [reply src/torrent.cpp]: I'll try helping with this.

The reason why I changed this line, is that I thought `!m_file_priority.empty()` expression could be redundant here. `std::find` will work for an empty container just as well.

Comment (d-komarov) [reply src/torrent.cpp]: If you don't like it, I'll revert this change easily.

Comment (arvidn) [reply src/torrent.cpp]: no, this is fine

File: src/torrent.cpp (hunk 6 above)
Comment (d-komarov) [reply src/torrent.cpp]: I'm not quite sure that this must be done, since `file_error_alert` and `torrent_error_alert` are not __critical__. That mean they could be lost and user won't have a chance to handle priority deviation.

Comment (d-komarov) [reply src/torrent.cpp]: I found a problem with this, `disk_io_thread::do_file_priority` always returns `no_error`, thus second branch will never be executed.

File: src/torrent.cpp (hunk 10 above)
Comment (d-komarov) [reply src/torrent.cpp]: I believe that most compilers will optimize this code, i.e. `fix_priorities` is about to execute only when `!m_torrent_file->num_pieces() || is_seed()` checks are passed.

However I don't mind moving those checks up.

File: test/test_torrent.cpp (hunk 12 above)
Comment (d-komarov) [reply test/test_torrent.cpp]: Good idea, but for now this should stay, isn't it?

File: src/torrent.cpp (hunk 6 above)
Comment (arvidn) [reply src/torrent.cpp]: oh, I think it was intended to propagate file errors.
Whether the user receives the alert or not is somewhat separate. I think it's important for libtorrent's internal state to stay consistent regardless.
The user can always ask for the priorities.

File: src/torrent.cpp (hunk 10 above)
Comment (arvidn) [reply src/torrent.cpp]: only if `fix_priorities()` and `torrent_file::files()` can be proven to be pure.

File: test/test_torrent.cpp (hunk 12 above)
Comment (arvidn) [reply test/test_torrent.cpp]: yeah

File: src/torrent.cpp [hunk 13]
```diff
@@ -5655,6 +5650,22 @@ namespace {
 
 	namespace
 	{
+		std::vector<boost::uint8_t> fix_priorities(std::vector<int> const& input, file_storage const& fs)
```
Comment (arvidn) [inline src/torrent.cpp]: factoring out this logic into a pure function is clearly a good thing. However, I'm a bit worried about merging this patch (correctly) into master. I'm a bit tempted to factor this function out in master instead of here, to keep this patch smaller

File: src/torrent.cpp [hunk 14]
```diff
@@ -5664,99 +5675,66 @@ namespace {
...
+			pause();
+		}
 	}
 
 	void torrent::prioritize_files(std::vector<int> const& files)
 	{
 		INVARIANT_CHECK;
 
-		// this call is only valid on torrents with metadata
-		if (!valid_metadata() || is_seed()) return;
```
Comment (arvidn) [inline src/torrent.cpp]: I take it it's a deliberate change to allow changing file priorities even when you're a seed. What's the rationale there? the problem of settings files to priority 0 once they're downloaded is that it may be very expensive to copy into the partfile.

File: src/torrent.cpp (hunk 13 above)
Comment (arvidn) [reply src/torrent.cpp]: I suppose this is required, because we need to convert `std::vector<int>` to `std::vector<std::uint8_t>` in order to make the disk thread call

File: src/torrent.cpp [hunk 15]
```diff
@@ -5664,99 +5675,66 @@ namespace {
...
 	}
 
 	void torrent::set_file_priority(int index, int prio)
 	{
-		INVARIANT_CHECK;
+		if (index < 0 || (valid_metadata() && index >= m_torrent_file->num_files())) { return; }
 
-		if (is_seed()) return;
+		// any unallocated slot is assumed to be 4
+		if (int(m_file_priority.size()) <= index && prio == 4) { return; }
```
Comment (arvidn) [inline src/torrent.cpp]: I don't this is safe anymore. there may be a set-priority-job in flight, that sets this file priority to something else. A subsequent set-priority must update it, even if it's called back-to-back

File: test/test_torrent.cpp [hunk 16]
```diff
@@ -248,63 +260,26 @@ TORRENT_TEST(added_peers)
...
 
-		// calculate the hash for all pieces
-		sha1_hash ph = hasher(&piece[0], piece.size()).final();
-		int num = t.num_pieces();
-		TEST_CHECK(t.num_pieces() > 0);
-		for (int i = 0; i < num; ++i)
-			t.set_hash(i, ph);
+	// calculate the hash for all pieces
+	TEST_CHECK(t.num_pieces() == 1);
+	t.set_hash(0, hasher(&piece[0], piece.size()).final());
```
Comment (arvidn) [inline test/test_torrent.cpp]: was it deliberate to only set the piece hash for the first piece? You made the torrent only have a single please. what's the rationale?

Comment (d-komarov) [reply test/test_torrent.cpp]: Previously there was only one piece too. Also I thought there was a mistake, torrent with only 1KB file and a 128 KB piece to calculate hash for. The last 127 KB of the piece should be padded with zeroes, and shouldn't participate in piece hash calculation. Does this make sense?

File: src/torrent.cpp (hunk 13 above)
Comment (d-komarov) [reply src/torrent.cpp]: Yes, also to reassign invalid priorities with correct values.

Comment (d-komarov) [reply src/torrent.cpp]: Hmm, will it make patch's size much smaller? However it's not a problem to move this code back to the `torrent::prioritize_files`.

Comment (d-komarov) [reply src/torrent.cpp]: I left some comments on it. Please take a look.

File: src/torrent.cpp (hunk 15 above)
Comment (d-komarov) [reply src/torrent.cpp]: This is true.

File: src/torrent.cpp (hunk 14 above)
Comment (d-komarov) [reply src/torrent.cpp]: This check was relocated to the line 5715.

**Comment by d-komarov** [comment]
Updated comments on https://github.com/d-komarov/libtorrent/pull/4

File: test/test_torrent.cpp [hunk 17]
```diff
@@ -285,18 +295,15 @@ TORRENT_TEST(torrent)
...
+		std::vector<char> piece(1024);
 		for (int i = 0; i < int(piece.size()); ++i)
 			piece[i] = (i % 26) + 'A';
 
 		// calculate the hash for all pieces
-		sha1_hash ph = hasher(&piece[0], piece.size()).final();
-		int num = t.num_pieces();
-		TEST_CHECK(t.num_pieces() > 0);
-		for (int i = 0; i < num; ++i)
-			t.set_hash(i, ph);
```
Comment (arvidn) [inline test/test_torrent.cpp]: I thought I reverted the changes to this test. As far as I can tell, the logic change does not require making these changes to the test (and it will be harder to merge in master). would you mind reverting these changes? (obviously leave in the new tests you added and the required changes like `wait_priority()`)

File: test/test_torrent.cpp [hunk 18]
```diff
@@ -81,48 +101,31 @@ void test_running_torrent(boost::shared_ptr<torrent_info> info, boost::int64_t f
...
 		TEST_EQUAL(st.total_wanted_done, 0);
-		if (!st.is_seeding)
-		{
-			TEST_EQUAL(int(h.file_priorities().size()), info->num_files());
-			TEST_EQUAL(h.file_priorities()[0], 0);
-			if (info->num_files() > 1)
-				TEST_EQUAL(h.file_priorities()[1], 0);
-			if (info->num_files() > 2)
-				TEST_EQUAL(h.file_priorities()[2], 1);
-		}
```
Comment (arvidn) [inline test/test_torrent.cpp]: I think these tests should still be here, or is there a reason to remove them?

**Comment by arvidn** [comment]
would you mind squashing these commits to?

**Comment by arvidn** [comment]
please include a not in the ChangeLog too

**Comment by d-komarov** [comment]
> please include a not in the ChangeLog too

I'm not quite sure what do you mean.

File: test/test_torrent.cpp (hunk 18 above)
Comment (d-komarov) [reply test/test_torrent.cpp]: Everything that is being tested here is now covered by `TEST_CHECK(prioritize_files(h, prio))`, plus it also checks the rest of priorities that should have been set.

**Comment by arvidn** [comment]
I meant to say "note". i.e. line describing the bug fix in the `ChangeLog` file

File: test/test_torrent.cpp [hunk 19]
```diff
@@ -285,16 +299,16 @@ TORRENT_TEST(torrent)
 		file_storage fs;
 
 		fs.add_file("test_torrent_dir2/tmp1", 1024);
-		libtorrent::create_torrent t(fs, 128 * 1024, 6);
+		libtorrent::create_torrent t(fs, 1024, 6);
```
Comment (arvidn) [inline test/test_torrent.cpp]: this used to be 128 blocks

File: test/test_torrent.cpp [hunk 20]
```diff
@@ -263,7 +278,6 @@ TORRENT_TEST(torrent)
 		std::vector<char> piece(128 * 1024);
 		for (int i = 0; i < int(piece.size()); ++i)
 			piece[i] = (i % 26) + 'A';
-
```
Comment (arvidn) [inline test/test_torrent.cpp]: please add back this blank line too. these tests have changed in master (I'm pretty sure) and this seems like an unnecessary conflict.

**Comment by d-komarov** [comment]
Oh, sure.

File: test/test_torrent.cpp (hunk 20 above)
Comment (d-komarov) [reply test/test_torrent.cpp]: This is commented code, that was removed in original PR. Are you sure it really make sense?

File: test/test_torrent.cpp (hunk 19 above)
Comment (d-komarov) [reply test/test_torrent.cpp]: The second argument is a `piece_size`. Setting it back to 128 * 1024 will require additional changes in `test_running_torrent()`.

File: test/test_torrent.cpp (hunk 20 above)
Comment (arvidn) [reply test/test_torrent.cpp]: it will still generate a conflict with master.

File: test/test_torrent.cpp [hunk 21]
```diff
@@ -52,10 +52,34 @@ POSSIBILITY OF SUCH DAMAGE.
...
+{
+	h.prioritize_files(prio);
+	return wait_priority(h, prio);
+}
+
 void test_running_torrent(boost::shared_ptr<torrent_info> info, boost::int64_t file_size)
 {
 	settings_pack pack = settings();
-	pack.set_int(settings_pack::alert_mask, alert::storage_notification);
+	pack.set_int(settings_pack::alert_mask, alert::progress_notification | alert::storage_notification);
```
Comment (arvidn) [inline test/test_torrent.cpp]: what's the `progress_notification` needed for?

File: test/test_torrent.cpp (hunk 19 above)
Comment (d-komarov) [reply test/test_torrent.cpp]: Alternative fix 13e92c31.

Comment (arvidn) [reply test/test_torrent.cpp]: A lot of this buffer juggling is a lot nicer in master, with `span`

Comment (d-komarov) [reply test/test_torrent.cpp]: The problem is not in particular buffer type, it is about how piece size for the single-file torrent is being obtained.

Comment (arvidn) [reply test/test_torrent.cpp]: I don't think I actually understand this (yet). Is this an independent problem from setting the file priorities?

Comment (d-komarov) [reply test/test_torrent.cpp]: From this test I discovered the problem for the first time. At that moment I didn't have a clue on what's causing test failure. So I began to investigate and had found some issues in it. When I fixed em', test still was failing randomly. Soon I found a root cause and it was file priorities, so I fixed it too.

TL;DR: Test issue appears to be independent, and could be submitted as a separate PR.

Comment (d-komarov) [reply test/test_torrent.cpp]: Without that changes test should be passing, but will take extra 10 seconds to finish. Also it is error prone, because of incorrect piece handling logic.

Comment (d-komarov) [reply test/test_torrent.cpp]: You decide, whether fixing test issues goes to this PR or another one. I just want this test be correct in the end.
